vnext/Desktop.IntegrationTests/HttpResourceIntegrationTests.cpp

```diff
@@ -354,6 +354,104 @@ TEST_CLASS (HttpResourceIntegrationTest) {
     Assert::AreEqual(static_cast<int64_t>(200), getResponse.StatusCode);
     Assert::AreEqual({"Redirect Content"}, content);
   }
+
+  TEST_METHOD(TimeoutSucceeds) {
+    auto port = s_port;
+    string url = "http://localhost:" + std::to_string(port);
+
+    promise<void> getPromise;
+    string error;
+    int statusCode = 0;
+
+    auto server = std::make_shared<HttpServer>(s_port);
+    server->Callbacks().OnGet = [&getPromise](const DynamicRequest &request) -> ResponseWrapper {
+      DynamicResponse response;
+      response.result(http::status::ok);
+
+      // Hold response to test client timeout
+      promise<void> timer;
+      timer.get_future().wait_for(std::chrono::milliseconds(2000));
+
+      return {std::move(response)};
+    };
+    server->Start();
+
+    auto resource = IHttpResource::Make();
+    resource->SetOnResponse([&getPromise, &statusCode](int64_t, IHttpResource::Response response) {
+      statusCode = static_cast<int>(response.StatusCode);
+      getPromise.set_value();
+    });
+    resource->SetOnError([&getPromise, &error, &server](int64_t, string &&errorMessage) {
+      error = std::move(errorMessage);
+      getPromise.set_value();
+    });
+    resource->SendRequest(
+        "GET",
+        std::move(url),
+        0, /*requestId*/
+        {}, /*headers*/
+        {}, /*data*/
+        "text", /*responseType*/
+        false, /*useIncrementalUpdates*/
+        6000, /*timeout*/
+        false, /*withCredentials*/
+        [](int64_t) {} /*callback*/);
+
+    getPromise.get_future().wait();
```

Comment:
>.wait(); [](http://example.com/codeflow?start=27&length=8)

With the code structured as is, are we waiting for the right thing here? Assuming this is to await the right moment to stop the server, wouldn't it be more succinct to wait for a signal that's set in the server OnGet lambda?

Reply:
A signal could work too.

However, we are not waiting just for the server to send a response, but for the client to receive it. (See `resource->SetOnResponse`).

Reply:
I meant "signal" in a generic way (including a promise). Ok, got you.